### DISCUSSION TITLE
idb explanation

### DISCUSSION THREAD

**Discussion by amityahav** [Hucaru/Valhalla#53] [Q&A]
Hey im going through this repo and i really wanna contribute to it. i do not understand fully he purpose of the idb file with all client functions. what are the use of them in the server perspective. as far as i understood packets are being sent back to the client and the client is responsible to decide which function to trigger based on the opcode it gets from the packet' header.
if i do want to contribute (lets say add the pin functionality) do i need something from the IDB file?

thanks

**Comment by Hucaru** [comment]
Hi, the idb files tells you what the packet structure for a particular opcode the client expects. This is useful as it tells us the structure of a message for a particular opcode the server needs to send to the client.

The idb will provide the opcode the server needs to send to the client for requesting a pin. Take a look at other servers that have the pin system implemented as it might make it more clear.

**Reply by amityahav** [reply]
Thanks for the fast reply I will look into it more deeply in the next couple of days.

**Comment by amityahav** [comment]
Hey, so after spending couple of hours understanding how to implement the PIN system, i have a question that i hope you could find a time to answer to:
while browsing a v.55/83 src code i've noticed that in order to make the client side display the PIN window box additional bytes need to be sent back to client on packetLoginResponce but i cant figure out what bytes and where in the packet structure, (see ref [here](https://github.com/ronancpl/HeavenMS/blob/master/src/tools/MaplePacketCreator.java#L716) also the packet structure is a bit diff than in 0.28) 
i've found an ugly way to pop up the PIN window by sending back to the client a packet with opcode 7 when returning from the handleLoginRequest (pin register) but this doesnt seem right to me, do you maybe know how can i find out what bytes and their positions? 
Thanks

**Reply by Hucaru** [reply]
I had a look in the idb and the pin request is a separate packet, you will therefore need to send a packet like:
```
func packetRequestPin() mpacket.Packet {
	pac := mpacket.CreateWithOpcode(0x07)
	pac.WriteByte(1) // 0 don't request pin, 1 register pin, 2 error with last pin entered re-enter, 3 general error message, 4 enter pin first time
	return pac
}
```

Somewhere here https://github.com/Hucaru/Valhalla/blob/master/login/handlers.go#L79 is where you will send this packet. You will then need to move the login success code to a new function for handling pin authorisation. There will need to also be logic for registering and retrieving pins as well as making this whole process dependent on config. I suggest get it working with sending a 0 for not requesting a pin and go from there.

**Reply by amityahav** [reply]
Thanks a lot!, i was thinking the same, could you elaborate how you find the different "modes" that come after the pin opcode?

**Reply by Hucaru** [reply]
The different modes are in the idb, I listed them above in the code

**Reply by amityahav** [reply]
yea i know, i was wondering how you found them in the IDB since its very hard to understand whats going on there.

**Reply by Hucaru** [reply]
Ah, If you follow the if statements in the idb you can tell there are 5 cases. Looking at the source example you linked you can see what the operations are:
```
        private static byte[] pinOperation(byte mode) {
                final MaplePacketLittleEndianWriter mplew = new MaplePacketLittleEndianWriter(3);
                mplew.writeShort(SendOpcode.CHECK_PINCODE.getValue());
                mplew.write(mode);
                return mplew.getPacket();
        }

        public static byte[] pinRegistered() {
                final MaplePacketLittleEndianWriter mplew = new MaplePacketLittleEndianWriter(3);
                mplew.writeShort(SendOpcode.UPDATE_PINCODE.getValue());
                mplew.write(0);
                return mplew.getPacket();
        }

        public static byte[] requestPin() {
                return pinOperation((byte) 4);
        }

        public static byte[] requestPinAfterFailure() {
                return pinOperation((byte) 2);
        }

        public static byte[] registerPin() {
                return pinOperation((byte) 1);
        }

        public static byte[] pinAccepted() {
                return pinOperation((byte) 0);
        }
```
